#1 move jailing code to core

IljaN opened the pull request:
<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please fill out below information carefully.

Please note that any kind of change first has to be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.
-->

## Description
- Read-only jailing of guest users is now provided by core
- Read-only groups can be configured in oc_appconfig under 'core' => 'read_only_groups' (json array)

## Related Issue
#https://github.com/owncloud/guests/issues/1

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->


## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

IljaN committed:
#1 move jailing code to core

lordelix committed:
Rename SHARE_TYPE_EMAIL to …_GUEST
Attach Share Plugin in OC instead of OCA

IljaN committed:
#1 Jail now allows initial creation of files dir

IljaN committed:
#16 Add possibility to exclude groups

PVince81 commented on settings/Controller/UsersController.php:
```diff
@@ -244,19 +246,42 @@ public function index($offset = 0, $limit = 10, $gid = '', $pattern = '', $backe
+				if (!empty($excludeGroups) && !empty($batch)) {
+					$excludeGroups = explode(';', $excludeGroups);
+					$excludedUsers = [];
+					foreach ($excludeGroups as $excludedGroup) {
```
so here you're post-filtering.

From what I see `displayNamesInGroup` was the function that did the filtering so maybe we should extend that one to include a "excludeGroups parameter" so we can optimize later on DB level.

Or we just leave the post-filtering if not too expensive...

IljaN replied:
@PVince81 Now that i see it, should i remove this code again? It is not need atm anyway, and displayNamesInGroup is probably the better place for it.

PVince81 replied:
yeah, keep it for another PR maybe

IljaN committed:
Default group "guests" for read_only_groups

IljaN committed:
default group was not set even with default value

IljaN committed:
Jail user based on guest flag in preferences

IljaN committed:
- Fixing share edit bug with read_only_jail
- Readding read_only_groups functionality

PVince81 commented on lib/private/Files/Storage/Wrapper/DirMask.php:
```diff
@@ -0,0 +1,190 @@
+
+use OC\Files\Storage\Wrapper\PermissionsMask;
+
+class DirMask extends PermissionsMask {
```
add PHPDoc, why do we need DirMask and how is it different from PermissionsMask ?

PVince81 commented on lib/private/Files/Storage/Wrapper/DirMask.php:
```diff
@@ -0,0 +1,190 @@
+	}
+
+	public function isSharable($path) {
+		if ($this->checkPath($path)) {
```
If we do decide to prevent sharing globally for guests, this is the place where we could always return `false`.

Let's leave this as is for now.

PVince81 commented on lib/private/Files/Storage/Wrapper/DirMask.php:
```diff
@@ -0,0 +1,190 @@
+	}
+
+	public function rename($path1, $path2) {
+		if (!$this->isReadable($path1)) {
```
does this need special handling for when $path1 is a part file and $path2 isn't ?

IljaN replied:
Why would it?

PVince81 replied:
I saw later on that you added the special handling in the subclass, ok

PVince81 commented on lib/private/Files/Storage/Wrapper/DirMask.php:
```diff
@@ -0,0 +1,190 @@
+			$storage = $this;
+		}
+		if ($this->checkPath($path)) {
+			return parent::getCache($path, $storage);
```
hmm, is this method override really needed ?

PVince81 commented on lib/private/User/Session.php:
```diff
@@ -395,8 +396,15 @@ protected function prepareUserLogin($firstTimeLogin) {
+				\OC_Util::copySkeleton($user, $userFolder);
+			} catch (NotPermittedException $ex) {
+				// possible if files directory is in an readonly jail
+				\OC::$server->getLogger()->info(
```
would "debug" be better here or "warning" ?

PVince81 commented on lib/private/legacy/util.php:
```diff
@@ -224,6 +224,51 @@ public static function setupFS($user = '') {
+			false
+		);
+
+		if ($isGuest === '1' || !empty($readOnlyGroupMemberships)) {
```
I suggest to only retrieve `$readOnlyGroupMemberships` and the related logic if `$isGuest` is true. Because in most cases the related logic does not need to run at all (and potentially slow down the code a few microseconds...)

IljaN replied:
You probably mean if $isGuest is false. if $isGuest is true we know for sure that we need the jail thus don`t need to concern ourselves with readonly groups.

PVince81 replied:
ah yes... I forgot that these are two separate mechanisms

PVince81 reviewed (commented):
Code looks good otherwise.

IljaN committed:
- Removing Group exclusion
- read_only_group code is only run if user is not a guest
- Change can`t create skeleton log message level to warning
- Describe dir mask
- Remove unneeded override inside dirmask

PVince81 commented:
Ready for rereview or anything remains to do ?

IljaN commented:
Yep ready

PVince81 commented:
Can we have some unit tests for the new classes ?

Integration tests would be nice too but can be done separately. Can you provide rough steps how to set this up for integration tests ?

PVince81 commented:
(the main code looks good)

IljaN commented:
There are two cases:
- User is guest by flag => create entry "isGuest=1" for the user in oc_preferences


- Create a group for example "guests" , add user to this group. Create a entry in oc_appconfig which looks like this: 
core	read_only_groups 	["guest_users"]

| appid | configkey | configvalue |
| ------------- | ------------- |-----------|
| core  | read_only_groups  | ["guests"] |

The outcome of both cases should be that the user is in a "Readonly Jail" and can`t upload files.
If a directory is shared with the user he can do everything in the folder according to the share permissions.

IljaN committed:
Added jailing unit-tests

PVince81 commented:
Looks good, thanks. 👍

mrow4a commented on lib/private/legacy/util.php:
```diff
@@ -224,6 +224,57 @@ public static function setupFS($user = '') {
+
+			$userGroups = array_keys(
+				\OC::$server->getGroupManager()->getUserIdGroups($user)
+			);
```
You are aware, that this introduced additional query - for each request going to oc, even most trivial? https://github.com/owncloud/core/issues/29380 @PVince81 @DeepDiver1975

mrow4a replied:
I mean, because isInGroup is calling exactly this function, it is not a problem (since it will be cached), but with central group table it might change (isInGroup should not fetch all groups of the users and then filter on app level)

mrow4a replied:
https://github.com/owncloud/core/pull/29384

PVince81 replied:
>  that this introduced additional query

Are you saying this was added recently and is a regression ? Or is it part of the years old code ? Can you git blame and link to the PR that introduced it ?
